src/coreclr/src/tools/Common/Sorting/MergeSortCore.cs

```diff
@@ -41,17 +41,15 @@ private static async Task ParallelSort(TDataStructure arrayToSort, int index, in
                 TDataStructureAccessor accessor = default(TDataStructureAccessor);
                 int halfLen = length / 2;
 
-                TaskCompletionSource<bool> rightSortComplete = new System.Threading.Tasks.TaskCompletionSource<bool>();
-                _ = Task.Run(async () =>
+                Task rightSortTask = Task.Run(async () =>
                 {
                     await ParallelSort(arrayToSort, index + halfLen, length - halfLen, comparer);
-                    rightSortComplete.SetResult(true);
```

Comment:
This change isn't safe. What happens is that this is converted into an Action delegate, which results in the task being considered finished before the actual sorting completes. however, I do see the bug. My task wizardy level is low, so what I would do to fix this is explicitly catch exceptions and if there is an exception set the TaskCompletionSource to have an exception in it instead of a result.

Reply:
@stephentoub What would be the best way to write this?

Reply:
Is the lambda really converted to an async void lambda? Doesn't this just use the `Func<Task>` overload, which should handle things correctly? (https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=netcore-3.1#System_Threading_Tasks_Task_Run_System_Func_System_Threading_Tasks_Task__)

Reply:
@jakobbotsch is correct.

This can just be:
```C#
Task rightSortTask = Task.Run(() => ParallelSort(...));
```

Reply:
Ah, when I hit this bug during development I possibly didn't have the code setup to return the Task correctly. Looking at it now, it appears to generate a use of `Task.Run(Func<Task>)` which I assume is correct. I think I may have set it up like...
```
Action parallelSort = async () => await ParallelSort (...);
Task.Run(parallelSort);

Instead of combining it all into 1 line which results in using the `Func<Action>` variant.